---
title: "Introduction to R: Bioconductor"
output: 
  md_document:
    variant: gfm
---

This final practical introduces Bioconductor and some basic analysis and annotations.

***

## Part 1: Bioconductor

[Bioconductor](https://www.bioconductor.org/) is a collection of more than 1,500 packages for analysing and annotating high-throughput genomic data.

It is useful for a wide range of applications including bulk and single-cell RNA sequencing, copy number analysis, flow cytometry, and methylation microarray data.

Before moving on, make sure to set your library path.

```{r eval=FALSE}
.libPaths("C:/fos_2019/library")
```

***

Unlike CRAN packages, Bioconductor packages are installed differently using `BiocManager`. You can check if the package you want to use is available:

```{r}
BiocManager::available("IRanges")
```

All Bioconductor packages have obligatory documentation, called 'vignettes'. You can view these from within R using:

```{r eval=FALSE}
browseVignettes("IRanges")
```

This takes you to a list of files that can help you when using this package in R. In addition to this, help is readily available on the [Bioconductor](https://support.bioconductor.org) forums.

Now load in the `IRanges` package.

```{r warning=FALSE, message=FALSE}
library(IRanges)
```

***

## Part 2: IRanges

This package provides efficient low-level and highly reusable classes for storing and manipulating **annotated ranges of integers**. It also contains useful functions, such as to identify overlaps.

You can imagine how this kind of functionality would be useful when handling sequences. `IRanges` forms the basis for one of the most important packages handling genomic data, `GenomicRanges`, which we will explore later.

`IRanges` objects can be created by specifying an integer range.

```{r}
ir <- IRanges(5,10)
ir
```

We have created an integer range from 5 to 10, and the width is listed as 6 since both the start and end integer contribute to the width of an `IRanges` object.

We can access this information using the `start()`, `end()`, and `width()` functions.

```{r}
start(ir)
end(ir)
width(ir)
```

***

`IRanges` can be shifted or resized. Here, `shift()` performs an intra range transformation and reduces the start and end integer by 2, keeping the width the same.

```{r}
shift(ir, -2)
```

The `resize()` function adjusts the width to a new specified value. Its default behaviour is to keep the `start` the same, and adjust the width using the `end` integer. 

```{r}
resize(ir, 3)
```

***

You can input multiple ranges at once using vectors to create a multi-range object.

```{r}
ir <- IRanges(c(4, 5, 8, 15, 19, 28, 40), width=c(15, 6, 7, 12, 9, 3, 6))
ir
```

These can also be resized or shifted, and this will be performed on each range individually.

```{r}
shift(ir,1)
```

***

The `reduce()` function will merge any overlapping regions and create an `IRanges` object of only distinct ranges. Here, there was a lot of overlap, with the only gap between 30 and 40.

```{r}
reduce(ir)
```

Any gaps can also be displayed easily. None of our `IRanges` object overlapped with integers 31 to 39.

```{r}
gaps(ir)
```

***

Each row in an `IRanges` object can be given a name, and they can be subsetted similarly to vectors or data frames.

```{r}
names(ir) = letters[1:7]
ir[c("b", "d", "e")]
```

Associating each range with a collection of attributes is very useful for genomic annotation. This metadata is assigned using the `mcols()`  function. 

R has some standard built-in data sets. One, called `mtcars` is used here to demonstrate annotation.

```{r}
mcols(ir) <- mtcars[1:7, 1:3]
ir
```

This annotation will remain the same through any `shift()` or `resize()` commands. Specific `mcol` variables can be accessed using the `$` operator, similar to in data frames.

There are over 300 functions that can be used to manipulate `IRanges` objects. In the next section, their role in genome-scale analysis will become clearer.

***

#### Question 1: Using the above created `IRanges` object:

* What is the mean `disp` for ranges whose `start` is less than 20?
* What is the maximum `width` for ranges whose `cyl` is 6?

***

## Part 2: Finding Overlaps

It is very useful to count overlaps in two distinct `IRanges` objects. We can subset the `ir` object above to create two new `IRanges` objects.

```{r}
ir1 <- ir[c(2,5,7)]
ir2 <- ir[-c(2,5,7)]
ir1
ir2
```

***

We can then use `findOverlaps()` to identify overlapping ranges in the `query` (first argument) and `subject` (second argument) objects.

```{r}
olaps <- findOverlaps(ir1, ir2)
olaps
```

The interpretation of the `findOverlaps()` output is as follows:

* The 1st range from `ir1` overlaps with the 1st range in `ir2` - [5,10] overlaps with [4,18]
* The 1st range from `ir1` overlaps with the 2nd range in `ir2` - [5,10] overlaps with [8,14]
* The 2nd range from `ir1` overlaps with the 3rd range in `ir2` - [19,27] overlaps with [15,26]

***

You can list the ranges with overlaps in a particular `IRanges` object after using `findOverlaps()` by utilising subsetting and `subjectHits()`.

So, we can show the ranges within `ir2` that overlap with ranges in `ir1`:

```{r}
ir2[subjectHits(olaps)]
```

***

You can also `countOverlaps()`.

```{r message=FALSE}
nolaps <- countOverlaps(ir2, ir1)
nolaps
```

So, here we see that `a`, `c`, and `d` ranges in `ir2` overlap with ranges in `ir1` once each, and `f` does not overlap with any ranges in `ir1`.

***

You can add output from `countOverlaps()` into your `IRanges`, coupling derived data with the original annotation.

```{r}
mcols(ir2)$Overlaps <- countOverlaps(ir2, ir1)
ir2
```

***

After this, you can perform co-ordinated actions, like subsetting for transcripts satisfying particular conditions.

```{r}
subset(ir2, Overlaps > 0)
```

***

#### Question 2: Load the example `IRanges` into R using the following command:

`load(url("https://raw.githubusercontent.com/molepi/Molecular-Data-Science/master/RIntro_practical/practical3_data_iranges.RData"))`

* How many overlaps are there between `ir1` and `ir2`?
* What is the name of the range in `ir2` with the most overlaps with `ir1`?
* Subset `ir2` to show only ranges that have more than 2 overlaps with `ir1`

***

## Part 4: Genomic Ranges

As an extension to `IRanges`, `GenomicRanges` objects contain obligatory metadata called `seqnames`. This gives the chromosome occupied by the sequence whose `start` and `end` position are modelled by the associated `IRanges`.

The package `Homo.sapiens` is an annotation package from Bioconductor. It contains information on genes in the human genome that can be accessed using the `genes()` function.

```{r warning=FALSE, message=FALSE}
library(Homo.sapiens)
hg <- genes(Homo.sapiens)
hg
```

This `GenomicRanges` object functions similarly to an `IRanges` object. You can explore `start()`, `end()`, and `width()` in the same way, and `subset()` it as before.

There are two parts to a `GenomicRanges` object. The first part is the `seqnames`, which must consist at least of the start and end coordinates along with the strand. The second part are additional elements or meta-data, like `GENEID`.

***

You can order a `GenomicRanges` object to make exploring it more intuitive. Here, we sort the ranges by chromosome using `seqnames()`.

```{r}
hg[order(seqnames(hg))]
```

***

An important component of `GenomicRanges`is `seqinfo()`. This shows how many base pairs are in each chromosome as `seqlengths` and which genome they from.

```{r}
seqinfo(hg)
```

***

You can inspect `seqinfo()` for a specific chromosome. The human X chromosome spans more than 155 million base pairs.

```{r}
seqinfo(hg)["chrX"]
```

***

You can `subset()` a `GenomicRanges` object to return only ranges in a specific chromosome, `seqnames()` alongside the `%in%` operator. Here, we see there are 296 genes on chromosome 21.

```{r}
subset(hg, seqnames %in% "chr21")
```

***

A useful function for `GenomicRanges` is the `keepStandardChromosomes()` function. This removes unmapped or non-standard chromosomes for the species of interest from the object.

```{r}
hg <- keepStandardChromosomes(hg, pruning.mode="coarse")
hg[order(seqnames(hg))]
```

If you compare this with the above output, you can see that ranges from chromosomes like `chrUn_gl000228` have been removed.

***

#### Question 3:

* How many base pairs does chromosome 15 span?
* Can you make a `table()` of the number of genes in all of the standard human chromosomes?

***

## Part 5: Overlaps with GWAS Hits

Genome-wide association studies (GWAS) are systematically represented in the EMBL-EBI GWAS catalog. The `gwascat` Bioconductor package allows us to retrieve a version of this.

```{r}
library(gwascat)
data(ebicat37)
ebicat37
```

Per GWAS finding, there is one range, and each corresponds to a SNP that has beem replicated as significantly associated with the given phenotype.

The genome here is listed as `GRCh37`, which is very similar to `hg19`. To avoid an error, we need to rename it.

```{r}
genome(ebicat37) <- "hg19"
```

***

You can view the top traits in the catalog using `topTraits()`.

```{r}
topTraits(ebicat37)
```

If we are interested in a specific trait, like LDL cholesterol, we can `subsetByTraits()`. 

```{r}
subsetByTraits(ebicat37, tr="LDL cholesterol")
```

***

We can use `findOverlaps()` as above to determine which genes have annotated intervals that overlap GWAS hits.

```{r}
goh <- findOverlaps(hg, ebicat37)
goh
```

To make more sense of this, we can count the number of genes that overlap with one or more GWAS hits using `length()` coupled with `unique()`.

```{r}
length(unique(queryHits(goh)))
```

***

Lastly, implementing the `reduce()` function, we can estimate the proportion of GWAS SNPs that lie within gene regions or, more specifically, exons.

```{r}
mean(reduce(ebicat37) %over% hg)
mean(reduce(ebicat37) %over% exons(Homo.sapiens))
```

We can see that the vast majority of GWAS hits lie in intergenic regions. This is the focus of a lot of research, since the regulatory mechanisms of non-coding regions is considerably complex and elusive.

***

## Feedback

Hopefully, this 2 day R course was informative and helpful. Please send any questions or comments to l.j.sinke@lumc.nl. 

Feedback is very much appreciated. Enjoy the rest of this Molecular Data Science FOS course! :)

***
