---
title: "Introduction to R"
output: 
  md_document:
    variant: markdown_github
---

* This practical is intended for students who have never used R before. Parts 1 and 2 are mandatory, Parts 3 and 4 are for fast students. At the end of this practical we expect you to be able to follow along with the other practicals of this course. Save your R-script or paste your answers in a Word document and mail it to k.f.dekkers@lumc.nl.

## Part 1: Explore R-studio & use the basics

* Open RStudio

* Open a new R-script with Ctrl-Shift-N, it should look like the following screenshot:
![Figure](https://github.com/molepi/Molecular-Data-Science/blob/master/RIntro_practical/figure.png?raw=true)

* You can type equations in the Console (see Figure) and R will calculate the result, try this by typing the following equation in the Console:

```{r}
1 + 1
```

* You can also type your code in the R-script (see Figure) and run a line or a selection using Ctrl-Enter. This is prefered to typing in the Console directly, since you can save the scripts for later use or share them with collaborators.

* Copy the following equations in your R-script, place your cursor on the first line and run the equations line by line. You can also select multiple lines and run them together.

```
5 - 2
2 * 2
6 / 2
3 ** 2
sqrt(9)
abs(-5)
```

* You can store values in variables for later use. You can view all your stored variables in the Environment (see Figure).

```{r}
x <- 1 + 1
x
```

* You can store multiple values in a vector using the concatenate function **c()**.

```{r}
y <- c(5, 6, 7)
y
```

* You can do calculus on stored variables, also on vectors.

```{r}
x + 1
x + y
y * y
```

* You can select, add and remove values from vectors.
* Tip: With **\#** you can comment your code so that other people can understand what you did. Comments are skipped when the code is run.

```{r}
y[1] # first value of y
y[2:3] # second and third value of y
c(y, 8) # add 8 to y
y[-3] # remove third value from y
```

* You can use functions on variables.

```{r}
sum(y)
max(y)
class(y)
str(y)
summary(y)
```

* If you want to know what the function does you can search for it in the Help pane (see Figure) or put a **?** before the function in your R-script or Console. Also, Google is often useful to get answers to R problems.

* Look at the Help page of some of the functions, either by searching for them or by running the following commands.

```
?sum
?class
```

* A vector can also contain characters.

```{r}
z <- c("a", "b", "c")
z
c(z, "d")
class(z)
```

* You can subset vectors in multiple ways:

```{r}
y[1] # first value of y
z[1:2] # first and second value of z
y[c(1, 3)] # first and third value of y
z[c(TRUE, TRUE, FALSE)] # first and second value of z
y[y == 6] # all values of y where y is 6
z[z != "a"] # all values of z where z is not "a"
y[y > 5] # all values of y where y is greater than 5
```

#### Question 1: Obtain the following vectors by adding, removing or subsetting vectors y or z. There are multiple ways to do this, but you only need to give one of them in your answers.

```
7
"a" "c"
5 7 9
"c" "b" "a"
```

* R also has the data frame class which you could compare to a spreadsheet.

* You can click on a data frame in the Environment (see Figure) to view it, however when the data frame is large it is better to use functions such as **str**, **summary**, **head** and **dim**.

```{r}
df <- data.frame(y, z) # create data frame from vectors y and z
df
class(df)
str(df)
summary(df)
```

* You can add row names and column names to data frames.

```{r}
rownames(df) <- c("row1", "row2", "row3")
colnames(df) <- c("col1", "col2")
df
```

* You can subset data frames.

```{r}
df[1, ] # first row of df
df[, 1] # first column of df
df[1, 1] # value at first row and first column of df
df$col1 # column col1
df[df$col1 > 5, ] # df where column col1 is greater than 5
df[df$col2 == "a", ] # df where column col2 is "a"
```

* You can add extra columns or rows to data frames.

```{r}
df$col3 <- c(TRUE, TRUE, FALSE) # add col3
df
df2 <- data.frame(col1 = 8, col2 = "d", col3 = FALSE) # create new data frame df2
rownames(df2) <- "row4" # add row name to df2
df2
df3 <- rbind(df, df2) # add row4 to df and store in variable df3
df3
```

#### Question 2:

1. Add a column with the name col4 to data frame df with the values 1 1 1.
2. Make a new data frame df4 which is a subset of data frame df but only contains columns col1 and col2 and rows row2 and row3.
3. Calculate the sum of col1 of data frame df3.
4. Change the row names of data frame df4 to "rowX" and "rowY".

## Part 2: Explore data set, plot and perform statistical tests

* We are now going to load a data set from the Framingham heart study [BioLINCC](biolincc.nhlbi.nih.gov).

```{r}
fhs <- read.csv(url("https://raw.githubusercontent.com/molepi/Molecular-Data-Science/master/RIntro_practical/data.csv"))
```

* First we will explore the data.

```{r}
str(fhs)
summary(fhs)
head(fhs)
sum(is.na(fhs))
```

* Description of variables at baseline: sex, total cholesterol, age, systolic blood pressure, diastolic blood pressure, current smoker, BMI, diabetis, blood pressure medication, glucose, education (four levels)

* Description of variables at follow up (24 years after baseline): myocardiac infarction

* There are missing values in the data. While there are different methods to handle missing data, we will in this case remove all individuals that have one or more missing values.

```{r}
fhs <- fhs[complete.cases(fhs), ]
sum(is.na(fhs))
```

* We can make some tables of the data.

```{r}
xtabs(~ MI, fhs)
prop.table(xtabs(~ MI, fhs))
xtabs(~ MI, fhs[fhs$SEX == "Female", ]) # table of MI for females only
```

#### Question 3:

1. Make a table of CURSMOKE
2. Make a proportion table of BPMEDS
3. Make a table of EDUC for individuals who smoke

* Packages can extent the functionality of R, **ggplot2** for example is a popular package for visualizations.

* The plots will appear in the Plots pane (see Figure), you can save the plots using the Export tab if you want.

```{r}
library(ggplot2)
ggplot(fhs, aes(MI)) + geom_bar()
ggplot(fhs, aes(BMI)) + geom_histogram(binwidth = 1)
ggplot(fhs[fhs$BMI > 30, ], aes(BMI)) + geom_density(fill = "black") 
```

* We can also make tables and plots of variable pairs.

```{r}
xtabs(~ MI + SEX, fhs)
ggplot(fhs, aes(x = SYSBP, y = DIABP)) + geom_point()
ggplot(fhs, aes(x = MI, y = AGE)) + geom_boxplot()
```

#### Question 4:

1. Make a bar plot of SEX
2. Make a histogram of TOTCHOL
3. Make a density plot of GLUCOSE for individuals with diabetes.
4. Make a table of MI AND EDUC for individuals that smoke.
5. Make a scatter plot of TOTCHOL and AGE
6. Make a box plot of MI and BMI for males.

* There are a lot of options to improve your plots. 
* Tip: **ggplot2** has a website with examples for all types of plots, see for example the page for [density plots](https://ggplot2.tidyverse.org/reference/geom_density.html).
* If you have time left you can explore some of these options and look at the ggplot2 website for inspiration.

```{r}
ggplot(fhs, aes(MI, fill = SEX)) + geom_bar(position = "dodge") + ggtitle("Bar plot") + 
  scale_fill_manual(values = c("blue4", "firebrick1")) + theme_bw()
ggplot(fhs, aes(BMI, fill = MI)) + geom_density(color = NA, alpha = 0.5) + ggtitle("Density plot") + 
  scale_fill_manual(values = c("blue4", "firebrick1")) + theme_bw()
ggplot(fhs, aes(x = SYSBP, y = DIABP, color = BPMEDS)) + geom_point() + 
  geom_smooth(method = "lm", se = F, color = "black") + ggtitle("Scatter plot") + 
  scale_color_manual(values = c("blue4", "firebrick1")) + theme_minimal()
ggplot(fhs, aes(x = MI, y = AGE, color = MI)) + geom_boxplot() + facet_wrap(~ SEX) + 
  ggtitle("Box plot") + scale_color_manual(values = c("blue4", "firebrick1")) + 
  theme_bw() + guides(color = F)
```

* Next we will perform some statistical tests.

```{r}
cor.test(~ SYSBP + DIABP, fhs) # is SYSBP correlated with DIABP
t.test(AGE ~ MI, fhs) # is AGE associated with MI (t-test)
fit <- glm(MI ~ AGE, fhs, family = "binomial") # is AGE associated with MI (logistic regression)
summary(fit)
```

#### Question 5:

1. Is AGE correlated with TOTCHOL?
2. Is MI associated with GLUCOSE (use t-test and logistic regression)?
3. Is MI associated with GLUCOSE in individuals with diabetes (use t-test and logistic regression) 

* One of the advantages of using R is to use loops to perform multiple analyses at once. The following loop calculates the association of MI with all the other variables and stores them in a data frame. We don't expect you to create loops like this in these 4 hours, but try to follow along with the code and see if it makes sense. In Part 3 we will explore some simple loops and self-made functions.

```{r}
results <- data.frame() # create empty data frame
for (i in colnames(fhs)[-ncol(fhs)]) { # loop over all column names except MI
  fit <- glm(MI ~ get(i), fhs, family = "binomial") # fit a model for the i-th column
  results <- rbind(results, summary(fit)$coefficients[2, , drop = FALSE]) # rbind second row of coefficients to results, first row is intercept
}
rownames(results) <- colnames(fhs)[-ncol(fhs)] # add row names
results$OR <- exp(results$Estimate) # add odds ratio
results[order(results$"Pr(>|z|)"), ] # order results by P-value
```

## Part 3: if-else statements, for loops and functions (optional)

* An if-statement is only run if the condition is true.

```{r}
if (1 < 2) { # if 1 smaller then 2
  print("1 is smaller than 2") # print "1 is smaller than 2"
}
```

* An else-statement is only run if the condition of the if-statement is false.

```{r}
if (1 > 2) {
  print("1 is greater than 2")
} else {
  print("1 is not greater than 2")
}
```

* With for-loops you assign a range of values (in the example values 1 to 10) to a variable (in the example variable i) and perform the actions in the loop for all these values in order (1 first, 10 last).

```{r}
for (i in 1:10) { # loop over values 1 to 10
  if (i < 5) {
    print(paste(i, "is smaller than 5"))
  } else {
    print(paste(i, "is not smaller than 5"))
  }
}
```

* You can also make your own functions in R.

```{r}
greater <- function(x, y) { # function requires arguments x and y
  if (x > y) {
    paste(x, "is greater than", y)
  } else {
    paste(x, "is not greater than", y)
  }
}
greater(1, 2)
greater(2, 1)
```

#### Question 6:

1. Make a for-loop, don't use subsetting like we practiced earlier, that returns all values of MI (of data frame fhs) where BMI is greater than 30.
2. Make a function that requires as arguments x, y and z and returns their product.
3. Make a loop that returns the association of all columns of data frame fhs with AGE, instead of using *glm* you can use *lm*.

## Part 4: Genomic ranges (optional)

* [Bioconductor](https://www.bioconductor.org/) contains a lot of packages to analyse biological data sets.

* The genomic ranges (?GRanges) object of the **GenomicRanges** package is often used in Bioconductor packages to store genomic data (SNP data, gene expression data, DNA methylation data, etc).

* Explore the following example of how to make and use genomic ranges. 

```{r}
library(GenomicRanges)

# make a genomic ranges object that resembles snp data
gr1 <- GRanges("chr1", IRanges(1:5, width = 1, names = paste0("snp", 1:5)), 
               ref_allele = sample(c("A", "G", "C", "T"), 5, replace = T), 
               alt_allele = sample(c("A", "G", "C", "T"), 5, replace = T)) 
gr1
gr1[1]
values(gr1)
ranges(gr1)
start(gr1)
end(gr1)
seqnames(gr1)
seqlevels(gr1)

# make a genomic ranges object that resembles gene expression data
gr2 <- GRanges("chr1", IRanges(4:8, width = 5,
                names = paste0("gene", 1:5)), GC_content = sample(1:100 / 100, 5))
gr2

# find snps in genes
hits <- findOverlaps(gr1, gr2)
hits
gr1[queryHits(hits)]
gr2[subjectHits(hits)]
data.frame(snps = names(gr1[queryHits(hits)]), genes = names(gr2[subjectHits(hits)]))

# make another genomic ranges object that resembles snp data
gr3 <- GRanges("chr1", IRanges(6:10, width = 1, names = paste0("snp", 6:10)), 
               ref_allele = sample(c("A", "G", "C", "T"), 5, replace = T), 
               alt_allele = sample(c("A", "G", "C", "T"), 5, replace = T))
gr3

# combine both snp data genomic ranges
gr4 <- c(gr1, gr3)
gr4

# find snps (of both snp data genomic ranges) in or near (within 2 base pairs) genes
hits <- findOverlaps(gr4, gr2, maxgap = 2)
hits
data.frame(snps = names(gr4[queryHits(hits)]), genes = names(gr2[subjectHits(hits)]))
```